Explain how docs:docsTest works (#24069)
gradle/gradle: subprojects/docs/README.md

It is possible to embed sample sources, commands, and expected output directly in the Asciidoc (or a mixture of embedded and `include`d), but we don't use this for the user manual yet. See the [Exemplar documentation](https://github.com/gradle/exemplar/#configuring-embedded-samples) if you're interested in this.

## Testing docs 

Currently, `docs` is tested by `docs:docsTest`, which covers three kinds of code: 

- The code generated by [Build Init Plugin](https://docs.gradle.org/current/userguide/build_init_plugin.html), i.e. `gradle init` task.

>   - `generate-samples.gradle.kts` registers multiple generator tasks that generate same sample project code as you manually run `gradle init`.

Reviewer: The samples generated using `init` are also published with the user manual. I think this should be mentioned because with the current text it looks like they are not.

Author: Added.

### `org.gradle.samples` plugin

`subprojects/docs/build.gradle` applies `generate-samples.gradle.kts`, which further applies an opinionated plugin called `org.gradle.samples`.
The source code of this plugin is [here](https://github.com/gradle/guides/blob/ba018cec535d90f75876bfcca29381d213a956cc/subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/internal/LegacySamplesDocumentationPlugin.java#L9).
This plugin adds a [`Samples`](https://github.com/gradle/guides/blob/fa335417efb5656e202e95759ebf8a4e60843f10/subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/Samples.java#L8) extension named `samples`.

This `samples` extension is configured in both `subprojects/docs/build.gradle` and `generate-samples.gradle.kts`:
all docs code to be tested will be assembled into [`samples.publishedSamples`](https://github.com/gradle/guides/blob/fa335417efb5656e202e95759ebf8a4e60843f10/subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/Samples.java#L41), as follows:

```

> This graph is generated by https://asciiflow.com/. You can copy-paste to the website to modify it. 

Author: This is how it looks like after being rendered:

<img width="925" alt="image" src="https://user-images.githubusercontent.com/12689835/221770083-7af7c41b-dfe5-432c-ab74-94068546f94a.png">

Reviewer 2: Do you think it makes sense to use mermaid instead of asciiflow, given that mermaid is supported natively by Github?

Author: Good to know!

After reading mermaid documentation, I'd still prefer the current ASCII way because:

- There is a little learning curve in mermaid, esp. for maintainers.
- It seems not easy to generate arbitrary graph.

Reviewer 3: 💭 
Regarding which tool to use. To me, the question is how far do we want to take the tool. If we want simple diagrams, then ASCII is the easiest entry, but Mermaid is a more powerful tool.  
We NEED more visuals in our docs. They should also look compelling.  I think this should wait for the tech writer to onboard for a few months, but I can see some of our complex diagrams, including flowcharts, sequence diagrams, and more, being replaced or created by Mermaid. 
In the meantime, ASCII is a simple use-case, and then we have future iterations of making visuals consistent.  IMO the hardest part is getting an idea of what the visual should be. Making it in the tool is easy by comparison.

It is possible to embed sample sources, commands, and expected output directly in the Asciidoc (or a mixture of embedded and `include`d), but we don't use this for the user manual yet. See the [Exemplar documentation](https://github.com/gradle/exemplar/#configuring-embedded-samples) if you're interested in this.

## Testing docs 

Currently, `docs` is tested by `docs:docsTest`, which covers three kinds of code: 

- The code generated by [Build Init Plugin](https://docs.gradle.org/current/userguide/build_init_plugin.html), i.e. `gradle init` task.

>   - `generate-samples.gradle.kts` registers multiple generator tasks that generate same sample project code as you manually run `gradle init`.

Reviewer 3: ```suggestion
  - `generate-samples.gradle.kts` registers multiple generator tasks that generate the same sample project code as you manually run `gradle init`.
```

  - These sample projects will also be published beside the user manual.
- The code samples under `subprojects/docs/src/samples`. They are published beside the user manual.
- [The code snippets](#code-snippets) under `subprojects/docs/src/snippets`, which are typically included in the user manual.

Note: the terminology `sample` could refer to different things depending on the context:

> - The code samples under `subprojects/docs/src/samples`. We'll call them "code sample".

Reviewer 3: ```suggestion
- The code samples under `subprojects/docs/src/samples`. We'll call them "code samples."
```

- The test unit used by [exemplar framework](https://github.com/gradle/exemplar/blob/9c4415f6237d7d86329fdbb32d80a2f7dd8ae0a9/samples-discovery/src/main/java/org/gradle/exemplar/model/Sample.java#L21). We'll call them "exemplar sample".

> The build script of `docs` subproject (`supprojects/docs/build.gradle`) eventually assembles the three kind of docs code above to exemplar samples then tests them with exemplar.

Reviewer 3: ```suggestion
The build script of the `docs` subproject (`supprojects/docs/build.gradle`) eventually assembles the three kinds of docs code above to exemplar samples and then tests them with exemplar.
```

### `org.gradle.samples` plugin

`subprojects/docs/build.gradle` applies `generate-samples.gradle.kts`, which further applies an opinionated plugin called `org.gradle.samples`.
The source code of this plugin is [here](https://github.com/gradle/guides/blob/ba018cec535d90f75876bfcca29381d213a956cc/subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/internal/LegacySamplesDocumentationPlugin.java#L9).
This plugin adds a [`Samples`](https://github.com/gradle/guides/blob/fa335417efb5656e202e95759ebf8a4e60843f10/subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/Samples.java#L8) extension named `samples`.

This `samples` extension is configured in both `subprojects/docs/build.gradle` and `generate-samples.gradle.kts`:
all docs code to be tested will be assembled into [`samples.publishedSamples`](https://github.com/gradle/guides/blob/fa335417efb5656e202e95759ebf8a4e60843f10/subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/Samples.java#L41), as follows:

```

> This graph is generated by https://asciiflow.com/. You can copy-paste to the website to modify it. 

Reviewer 3: 💄 
```suggestion
This graph is generated by [asciiflow.com](https://asciiflow.com/). You can copy-paste it to the website to modify it. 
```
